### PR TITLE
Implement `Option::take_if`

### PR DESCRIPTION
**Pull request by kellerkindt**
Tracking issue: #98934
ACP: rust-lang/libs-team#70 [accepted]

### CODE REVIEW

**Comment by kellerkindt** [comment]
Hey @Mark-Simulacrum, [further input needed for the trait bounds](https://github.com/rust-lang/rust/issues/98934#issuecomment-1177278646). I copied them over from the `Option::filter` impl, but I am not familiar with the whole `~const` thing. Could you thake a look? :)

**Comment by Mark-Simulacrum** [comment]
Hi! Per the rustbot comment (https://github.com/rust-lang/rust/pull/98935#issuecomment-1175108127), new APIs should be proposed as a T-libs-api API change proposal -- has one been filed for this? Can you link to it?

With regards to the const bounds, my recommendation is to drop the const_unstable + const part of this function for now; the story around `~const` is likely to shift significantly and isn't near stabilization; we don't need to worry about it for new APIs being added. (It seems to compile, based on CI, so in that case it's likely fine to leave them as-is; not clear if you're indicating there's some problem...).

**Comment by Mark-Simulacrum** [comment]
The tracking issue looks pretty good! However, typically our process is that we:

* [File an ACP issue](https://github.com/rust-lang/libs-team/issues/new?assignees=&labels=api-change-proposal%2C+T-libs-api&template=api-change-proposal.md&title=%28My+API+Change+Proposal%29), pretty much with the text you've put in the tracking issue
* In parallel with that (or after), file a PR linking to the proposal issue and file a tracking issue for the feature
* Then, the PR gets reviewed and merged

You definitely don't need to write a full RFC for this -- just needs that issue on the libs-team repo so that they can review and approve the change going in.

**Comment by kellerkindt** [comment]
> You definitely don't need to write a full RFC for this -- just needs that issue on the libs-team repo so that they can review and approve the change going in.

Thanks for clarification :)
ACP: rust-lang/libs-team#70

**Comment by JohnCSimon** [comment]
still waiting on ACP.

**Comment by JohnCSimon** [comment]
@kellerkindt I guess this this is still waiting on ACP?

**Comment by kellerkindt** [comment]
Yeah, there seems to be movement? Or did I miss something to start the actual discussion?

**Comment by JohnCSimon** [comment]
> Yeah, there seems to be no movement? Or did I miss something to start the actual discussion?

I don't know, sorry. Maybe ask people on zulip  - https://rust-lang.zulipchat.com/
like `@the8472` who commented here https://github.com/rust-lang/libs-team/issues/70#issuecomment-1194554228

File: library/core/src/option.rs [hunk 1]
```diff
@@ -1305,6 +1305,46 @@ impl<T> Option<T> {
...
+    /// });
+    /// assert_eq!(x, Some(43));
+    ///
+    /// x.retain(|v| *v == 42);
+    /// assert_eq!(x, None);
+    /// ```
+    #[inline]
+    #[unstable(feature = "option_retain", issue = "98934")]
+    #[rustc_const_unstable(feature = "const_option_retain", issue = "98934")]
+    pub const fn retain<P>(&mut self, predicate: P)
```
Comment (Mark-Simulacrum) [inline library/core/src/option.rs:1335]: It looks like the ACP was accepted as take_if, not retain? Can you update accordingly?

**Pull request merged**